bundles/org.openhab.binding.tsmart/src/main/java/org/openhab/binding/tsmart/internal/TSmartUDPListener.java

```diff
@@ -0,0 +1,143 @@
+     * This method adds the address and ThingHandler to the map.
+     *
+     * @param addr Network address of device
+     * @param handler Instance of ThingHander
+     */
+    public static void addHandler(InetAddress addr, TSmartHandler handler) {
+        if (handlerAddressMap.containsKey(addr)) {
+            // throw error
+        } else {
+            handlerAddressMap.put(addr, handler);
+            enableListener();
+        }
+    }
+
+    /**
+     * Maintain a mapping between a device address and the openHAB handler
+     * in order to route any received UDP packets to the correct handler.
+     * This method removes the mapping.
+     *
+     * @param addr Network address of device
+     */
+    public static void removeHandler(InetAddress addr) {
+        if (handlerAddressMap.containsKey(addr)) {
+            handlerAddressMap.remove(addr);
+        }
+    }
+
+    /**
+     * Add the DiscoveryService to handle any UDP packets as responses
+     * to discovery broadcast packets.
+     *
+     * @param discoService Instance of the TSmartDiscoveryService
+     */
+    public static void startDiscovery(TSmartDiscoveryService discoService) {
+        TSmartUDPListener.discoService = discoService;
+        enableListener();
+    }
+
+    /**
+     * Stop listening for responses to discovery broadcast packets
+     */
+    public static void stopDiscovery() {
+        discoService = null;
+    }
+
+    /**
+     * Method to receive UDP packets and route them to the appropriate
+     * ThingHandler or DiscoveryService.
+     */
+    @Override
+    public void run() {
+
+        try {
+            DatagramSocket socket = new DatagramSocket(T_SMART_PORT);
+
+            while (isEnabled()) {
+
+                byte[] buffer = new byte[512];
+                DatagramPacket response = new DatagramPacket(buffer, buffer.length);
+                socket.receive(response);
+                if (handlerAddressMap.containsKey(response.getAddress())) {
+
+                    TSmartHandler handler = handlerAddressMap.get(response.getAddress());
+
+                    if (handler != null && buffer[0] == (byte) 0xF1) {
+                        handler.updateStatusHandler(buffer);
+                    }
+                } else {
+                    TSmartDiscoveryService discoService = TSmartUDPListener.discoService;
+                    if (discoService != null && buffer[0] == (byte) 0x01 && buffer[3] != (byte) 0x54) {
+                        discoService.handleDiscoveryResponse(response.getAddress(), buffer);
+                    }
+                }
+            }
+            socket.close();
+            listener = null;
+            discoService = null;
+        } catch (SocketTimeoutException ex) {
+            logger.debug("Timeout error: {}", ex.getMessage());
+        } catch (IOException ex) {
+            logger.debug("Client error: {}", ex.getMessage());
+        }
```

Comment:
Should this be propagated back to the handlers, so they could take appropriate actions? For example, a thing handler might want to set status to offline with communication error.

Reply:
Agree - these should both just mark all Things as offline because the UDP listener has failed. Will implement